qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -161,6 +167,7 @@ def cost_operator(self, cost_operator) -> None:
             cost_operator (OperatorBase, optional): cost operator to set.
         """
         self._cost_operator = cost_operator
+        self.qregs = [QuantumRegister(self.num_qubits, name="q")]
```

Comment:
In QAOA, the size of the circuit is directly specified by the size of the cost operator. It's there so if users change the dimension of the cost operator the number of qubits is adjusted, too. 

Yeah the workflow you show above is confusing, I agree! But I think that's a deeper problem: we should ensure blueprint circuits can only be modified in the intended ways, and e.g. adding registers or gates would either be disallowed or return a circuit that's not a blueprint anymore. Because whatever modification you do will get wiped when the blueprint circuit rebuilds itself 😄 E.g.
```python
>>> from qiskit.circuit.library import RealAmplitudes
>>> circuit = RealAmplitudes(1, reps=1)
>>> circuit.h(0)
<qiskit.circuit.instructionset.InstructionSet object at 0x7fb3e95039d0>
>>> circuit.draw()
     ┌───────────────────────────┐┌───┐
q_0: ┤ RealAmplitudes(θ[0],θ[1]) ├┤ H ├
     └───────────────────────────┘└───┘
>>> circuit.num_qubits = 3
>>> circuit.draw()
     ┌────────────────────────────────────────────────┐
q_0: ┤0                                               ├
     │                                                │
q_1: ┤1 RealAmplitudes(θ[0],θ[1],θ[2],θ[3],θ[4],θ[5]) ├
     │                                                │
q_2: ┤2                                               ├
     └────────────────────────────────────────────────┘
>>>

```

Reply:
Yeah, I think you're right.  This is definitely a bigger issue than this PR - let's make a separate issue for it, and handle it for 0.19.  We can't fix it in 0.18.1 anyway, because it'd be backwards incompatible.